src/routes/events/containers/EventsPage.js

```diff
@@ -6,41 +6,85 @@ import Spinner from 'react-spinkit'
 import withFetcher from 'modules/endpoint/hocs/withFetcher'
-import withAvailableLanguageUpdater from 'modules/language/hocs/withAvailableLanguageUpdater'
-import EventDetail from './EventDetail'
+import EventDetail from '../components/EventDetail'
 import EventList from '../components/EventList'
+import { setLanguageChangeUrls } from 'modules/language/actions/setLanguageChangeUrls'
+import LanguageModel from 'modules/endpoint/models/LanguageModel'
 
 /**
  * Displays a list of events or a single event, matching the route /<location>/<language>/events*
  */
-class EventsPage extends React.Component {
+export class EventsPage extends React.Component {
   static propTypes = {
     events: PropTypes.arrayOf(PropTypes.instanceOf(EventModel)).isRequired,
     location: PropTypes.string.isRequired,
+    languages: PropTypes.arrayOf(PropTypes.instanceOf(LanguageModel)).isRequired,
     language: PropTypes.string.isRequired,
+    dispatchLanguageChangeUrls: PropTypes.func.isRequired,
     path: PropTypes.string
   }
 
   getPath () {
     return `/${this.props.location}/${this.props.language}/events`
   }
 
-  render () {
-    let events = this.props.events.map((event, index) => ({event: event, thumbnailPlaceholder: index % 3}))
+  mapLanguageToUrl = (language, id) =>
+    id ? `/${this.props.location}/${language}/events/${id}`
+      : `/${this.props.location}/${language}/events`
+
+  componentDidMount () {
+    // all events
+    let availableLanguages = {}
+
+    if (this.props.path && this.props.events) {
+      // specific event
+      const event = this.props.events.find(
+        (event) => event.id.toString() === this.props.path.replace('/', '')
+      )
+      if (event) availableLanguages = event.availableLanguages
+    }
+    this.props.dispatchLanguageChangeUrls(this.mapLanguageToUrl, this.props.languages, availableLanguages)
+  }
+
+  // we must not call dispatch in componentWillUpdate or componentDidUpdate
+  componentWillReceiveProps (nextProps) {
+    if (nextProps.events === this.props.events && nextProps.path === this.props.path) {
+      // no relevant prop changes
+      return
+    }
 
+    if (nextProps.path) {
+      // specific event
+      const event = nextProps.events.find(
+        (event) => event.id.toString() === nextProps.path.replace('/', '')
+      )
+      if (event) {
+        // events have been loaded in the new language
+        this.props.dispatchLanguageChangeUrls(
+          this.mapLanguageToUrl, nextProps.languages, event.availableLanguages
+        )
+      }
+    } else {
+      // all events
+      this.props.dispatchLanguageChangeUrls(this.mapLanguageToUrl, nextProps.languages)
+    }
+  }
+
+  render () {
     if (this.props.path) {
       // event with the given id from this.props.path
-      const event = events.find((event) => event.event.id.toString() === this.props.path.replace('/', ''))
+      const event = this.props.events.find((event) => event.id.toString() === this.props.path.replace('/', ''))
```

Comment:
You have this code 3 times in this file: `this.props.events.find((event) => event.id.toString() === this.props.path.replace('/', ''))`
Better extract it to a method

Reply:
done